### ISSUE TITLE
[Bug]: AssertionError in sampler.py:383

### ISSUE THREAD

**Issue by Mamohan43** [vllm-project/vllm#41031]
### Your current environment

**Summary:**
During boundary fuzzing of the vLLM inference server (SDK AIC.1.22.0.62) using the SentinelFuzz tool, we identified a critical stability bug in the sampler component. Sending a single API request with prompt_logprobs=20 causes an AssertionError at sampler.py:383, which terminates the server process. All in-flight requests are killed. The server requires a manual restart to recover. This bug is present in open-source vLLM code and is reproducible 100% of the time.

**File & Location**
Repository: vllm-project/vllm (open source)
File: vllm/model_executor/layers/sampler.py
Line: 383
Error: AssertionError (assertion in logits accounting loop)
Stack trace class: PROCESS_CRASH (no HTTP response — connection refused after crash)

**How We Found It**
This bug was discovered by SentinelFuzz, an boundary fuzzing tool that generates semantically-aware adversarial API requests targeting known parameter boundaries in the LLM inference pipeline. The tool's knowledge base identified prompt_logprobs as a parameter that interacts with the sampler's internal logits accounting loop and generated boundary-value test cases above the normally tested range.

1 |Start the vLLM inference server with model meta-llama/Llama-3.1-8B on in standard (non-disaggregated) mode.
2 Confirm the server is healthy: send a normal request with prompt_logprobs=1 and verify HTTP 200 with token output.
3 Send the adversarial request below with prompt_logprobs=20.
4 Observed results: Connection refused. Server process has exited. 

Server logs show AssertionError at sampler.py:383. All subsequent requests fail until manual restart.

Reproducer — Bug 1: sampler.py:383 AssertionError
Model: meta-llama/Llama-3.1-8B

curl -X POST http://localhost:8000/v1/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer <your-api-key>" \
  -d '{
      "model": "meta-llama/Llama-3.1-8B",
      "prompt": "Hello world",
      "max_tokens": 10,
      "temperature": 0.0,
      "prompt_logprobs": 20  # <— TRIGGER: value above safe boundary
    }'

Expected (normal): HTTP 200 with token output
Actual: Connection refused. Server process has exited.
Server log: AssertionError at sampler.py:383

### 🐛 Describe the bug

To fix this issue, the solution is to add `None` guards to the metric computations in `UsageInfo` construction, defaulting to `0.0` when either operand is `None`. This is a minor, defensive change that has no impact on normal requests. Note that `arrival_time` is always set, so errors only occur when `finished_time` or `first_scheduled_time` are `None`, indicating the scheduler is not setting these values for certain requests.

def _safe_diff_ms(a, b):
    return (a - b) * 1000 if (a is not None and b is not None) else 0.0

usage = UsageInfo(
    prompt_tokens=num_prompt_tokens,
    completion_tokens=num_generated_tokens,
    total_tokens=num_prompt_tokens + num_generated_tokens,
    ttft_excluding_queue_wait_time_in_ms=_safe_diff_ms(
        final_res.metrics.first_token_time,
        final_res.metrics.first_scheduled_time),
    e2e_inference_in_ms=_safe_diff_ms(
        final_res.metrics.finished_time,
        final_res.metrics.arrival_time),
    queue_wait_time_in_ms=_safe_diff_ms(
        final_res.metrics.first_scheduled_time,
        final_res.metrics.arrival_time),
    ttft_in_ms=_safe_diff_ms(
        final_res.metrics.first_token_time,
        final_res.metrics.arrival_time),
)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Mamohan43** [comment]
To fix this issue, the solution is to add None guards to the metric computations in UsageInfo construction, defaulting to 0.0 when either operand is None. This is a minor, defensive change that has no impact on normal requests. Note that arrival_time is always set, so errors only occur when finished_time or first_scheduled_time are None, indicating the scheduler is not setting these values for certain requests.

def _safe_diff_ms(a, b):
return (a - b) * 1000 if (a is not None and b is not None) else 0.0

usage = UsageInfo(
prompt_tokens=num_prompt_tokens,
completion_tokens=num_generated_tokens,
total_tokens=num_prompt_tokens + num_generated_tokens,
ttft_excluding_queue_wait_time_in_ms=_safe_diff_ms(
final_res.metrics.first_token_time,
final_res.metrics.first_scheduled_time),
e2e_inference_in_ms=_safe_diff_ms(
final_res.metrics.finished_time,
final_res.metrics.arrival_time),
queue_wait_time_in_ms=_safe_diff_ms(
final_res.metrics.first_scheduled_time,
final_res.metrics.arrival_time),
ttft_in_ms=_safe_diff_ms(
final_res.metrics.first_token_time,
final_res.metrics.arrival_time),
)
